builtin/logical/database/path_rotate_credentials.go

```diff
@@ -78,22 +78,30 @@ func (b *databaseBackend) pathRotateRootCredentialsUpdate() framework.OperationF
 			return nil, err
 		}
 
-		// Take out the backend lock since we are swapping out the connection
-		b.Lock()
-		defer b.Unlock()
-
 		// Take the write lock on the instance
 		dbi.Lock()
-		defer dbi.Unlock()
-
+		unlocked := false
+		defer func() {
+			if !unlocked {
+				// set to true before unlocking to avoid race condition
+				unlocked = true
+				dbi.Unlock()
+			}
+		}()
 		defer func() {
 			// Close the plugin
 			dbi.closed = true
 			if err := dbi.database.Close(); err != nil {
 				b.Logger().Error("error closing the database plugin connection", "err", err)
 			}
 			// Even on error, still remove the connection
-			delete(b.connections, name)
+			// be careful with a deadlock here
+			if !unlocked {
+				// set to true before unlocking to avoid race condition
+				unlocked = true
+				dbi.Unlock()
+			}
+			b.ClearConnection(name)
```

Comment:
I think that this can be racy if another request gets to grab the write lock between this unlock and clear connection, potentially clearing out a newly set connection.

Reply:
Good call. I'll add a method to only clear the connection if the ID matches.